File: ctrl/core/picker.js
import { cancelled, failure, ResultStatus } from "./results.js";

const CANCEL_TOKENS = new Set(["q", "quit", "cancel", "/cancel"]);

function normalized(value) {
  return String(value ?? "").trim().toLocaleLowerCase();
}

function isCancellation(value) {
  return CANCEL_TOKENS.has(normalized(value));
}

function scoreDescriptor(descriptor, query) {
  if (query === "") return 1;
  const id = normalized(descriptor.id);
  const title = normalized(descriptor.title);
  const description = normalized(descriptor.description);
  if (id === query || title === query) return 100;
  if (id.startsWith(query) || title.startsWith(query)) return 80;
  if (id.includes(query) || title.includes(query)) return 60;
  if (description.includes(query)) return 20;
  return 0;
}

export function searchActionDescriptors(descriptors, query = "") {
  const needle = normalized(query);
  return descriptors
    .map((descriptor) => ({ descriptor, score: scoreDescriptor(descriptor, needle) }))
    .filter(({ score }) => score > 0)
    .sort((a, b) => b.score - a.score || a.descriptor.id.localeCompare(b.descriptor.id))
    .map(({ descriptor }) => descriptor);
}

function selectedOption(options, rawValue) {
  const numeric = Number.parseInt(String(rawValue).trim(), 10);
  if (!Number.isInteger(numeric) || numeric < 1 || numeric > options.length) return undefined;
  return options[numeric - 1];
}

function renderOptions(write, options, label) {
  write(label);
  options.forEach((option, index) => write(`${index + 1}. ${option.label}`));
}

async function resolveSelectableValues(input, registry, context) {
  if (Array.isArray(input.choices)) {
    return { ok: true, values: input.choices.map((value) => ({ label: String(value), value })) };
  }

  const source = input.selectionAction ?? input.selectableSource;
  if (!source) return { ok: true, values: null };
  if (typeof source.action !== "string" || source.action === "") {
    return {
      ok: false,
      result: failure(null, ResultStatus.INTERNAL_FAILURE, `Selectable input ${input.name} does not name the canonical Action used to resolve its values.`),
    };
  }

  const result = await registry.execute(source.action, source.input ?? {}, context);
  if (!result.ok) return { ok: false, result };
  const collection = result.data?.[source.collection];
  if (!Array.isArray(collection)) {
    return {
      ok: false,
      result: failure(source.action, ResultStatus.INTERNAL_FAILURE, `Selectable source ${source.action} did not return ${source.collection}.`),
    };
  }
  const valueField = source.valueField;
  return {
    ok: true,
    values: collection.map((item) => {
      const value = valueField ? item?.[valueField] : item;
      return { label: String(value), value };
    }),
  };
}

export async function runGuidedActionPicker({ registry, context = {}, prompt, write = () => {} }) {
  if (!registry || typeof registry.list !== "function" || typeof registry.execute !== "function") {
    throw new TypeError("Guided picker requires an Action registry.");
  }
  if (typeof prompt !== "function") throw new TypeError("Guided picker requires prompt().");
  if (typeof write !== "function") throw new TypeError("Guided picker write must be a function.");

  const query = await prompt("Search Actions (blank for all, q to cancel): ");
  if (isCancellation(query)) return cancelled(null, "Action selection cancelled.");

  const matches = searchActionDescriptors(registry.list(), query);
  if (matches.length === 0) {
    return failure(null, ResultStatus.INVALID_INPUT, `No Actions match: ${String(query).trim()}`);
  }

  const actionOptions = matches.map((descriptor) => ({ label: `${descriptor.id} — ${descriptor.title}`, value: descriptor }));
  renderOptions(write, actionOptions, "Actions:");
  const rawActionChoice = await prompt("Select Action number (q to cancel): ");
  if (isCancellation(rawActionChoice)) return cancelled(null, "Action selection cancelled.");
  const selected = selectedOption(actionOptions, rawActionChoice);
  if (!selected) return failure(null, ResultStatus.INVALID_INPUT, "Action selection must be one of the displayed numbers.");

  const action = selected.value;
  const inputValues = {};
  for (const input of action.inputs ?? []) {
    const resolved = await resolveSelectableValues(input, registry, context);
    if (!resolved.ok) return resolved.result;

    if (resolved.values) {
      if (resolved.values.length === 0 && input.required) {
        return failure(action.id, ResultStatus.INVALID_INPUT, `No selectable values are available for ${input.name}.`);
      }
      if (resolved.values.length > 0) {
        renderOptions(write, resolved.values, `${input.name}:`);
        const rawChoice = await prompt(`Select ${input.name} number (q to cancel): `);
        if (isCancellation(rawChoice)) return cancelled(action.id, `${action.title} cancelled.`);
        const chosen = selectedOption(resolved.values, rawChoice);
        if (!chosen) return failure(action.id, ResultStatus.INVALID_INPUT, `${input.name} selection must be one of the displayed numbers.`);
        inputValues[input.name] = chosen.value;
        continue;
      }
    }

    const rawValue = await prompt(`${input.name}${input.required ? "" : " (optional)"} (q to cancel): `);
    if (isCancellation(rawValue)) return cancelled(action.id, `${action.title} cancelled.`);
    if (input.required && String(rawValue).trim() === "") {
      return failure(action.id, ResultStatus.INVALID_INPUT, `${input.name} is required.`);
    }
    if (String(rawValue) !== "" || input.required) inputValues[input.name] = rawValue;
  }

  return registry.execute(action.id, inputValues, context);
}

File: ctrl/src/picker.rs
use crate::action::{
    ActionDescriptor, ActionExecutionContext, ActionInputDefinition, ActionRegistry, MutationClass,
};
use crate::result::ActionResult;
use serde_json::{Map, Value};
use std::cmp::Ordering;
use std::io::{self, Write};

pub trait TerminalSurface {
    fn write_line(&mut self, line: &str);
    fn prompt(&mut self, message: &str) -> Option<String>;
}

#[derive(Default)]
pub struct StdioTerminalSurface;

impl TerminalSurface for StdioTerminalSurface {
    fn write_line(&mut self, line: &str) {
        eprintln!("{line}");
    }

    fn prompt(&mut self, message: &str) -> Option<String> {
        eprint!("{message}");
        let _ = io::stderr().flush();
        let mut input = String::new();
        match io::stdin().read_line(&mut input) {
            Ok(0) | Err(_) => None,
            Ok(_) => Some(input.trim_end().to_owned()),
        }
    }
}

#[derive(Default)]
pub struct NullTerminalSurface;

impl TerminalSurface for NullTerminalSurface {
    fn write_line(&mut self, _line: &str) {}

    fn prompt(&mut self, _message: &str) -> Option<String> {
        None
    }
}

enum Flow<T> {
    Value(T),
    Back,
    Cancelled(ActionResult),
}

fn cancel(action: Option<&str>) -> ActionResult {
    ActionResult::cancelled(action, "Guided Action selection cancelled.")
}

fn action_score(action: &ActionDescriptor, query: &str) -> Option<u8> {
    if !action.availability.available {
        return None;
    }
    let query = query.trim().to_lowercase();
    if query.is_empty() {
        return Some(5);
    }
    let id = action.id.to_lowercase();
    let title = action.title.to_lowercase();
    let description = action.description.to_lowercase();
    if id == query {
        Some(0)
    } else if id.starts_with(&query) {
        Some(1)
    } else if id.contains(&query) {
        Some(2)
    } else if title.contains(&query) {
        Some(3)
    } else if description.contains(&query) {
        Some(4)
    } else {
        None
    }
}

pub fn search_action_descriptors(registry: &ActionRegistry, query: &str) -> Vec<ActionDescriptor> {
    let mut actions = registry
        .list()
        .into_iter()
        .filter_map(|action| action_score(&action, query).map(|score| (score, action)))
        .collect::<Vec<_>>();
    actions.sort_by(|(left_score, left), (right_score, right)| {
        left_score
            .cmp(right_score)
            .then_with(|| left.id.cmp(&right.id))
            .then(Ordering::Equal)
    });
    actions.into_iter().map(|(_, action)| action).collect()
}

fn select_action(registry: &ActionRegistry, surface: &mut dyn TerminalSurface) -> Flow<ActionDescriptor> {
    loop {
        let Some(raw_query) = surface.prompt("Search Actions (/cancel): ") else {
            return Flow::Cancelled(cancel(None));
        };
        if raw_query.trim() == "/cancel" {
            return Flow::Cancelled(cancel(None));
        }
        let actions = search_action_descriptors(registry, &raw_query);
        if actions.is_empty() {
            surface.write_line("No Actions match that search.");
            continue;
        }
        surface.write_line("Matching Actions:");
        for (index, action) in actions.iter().enumerate() {
            surface.write_line(&format!(
                "{}. {} — {} [{}]",
                index + 1,
                action.id,
                action.title,
                action.mutation_class.as_str(),
            ));
        }

        loop {
            let Some(raw) = surface.prompt("Select Action by number or id (/back, /cancel): ") else {
                return Flow::Cancelled(cancel(None));
            };
            let choice = raw.trim();
            if choice == "/cancel" {
                return Flow::Cancelled(cancel(None));
            }
            if choice == "/back" {
                break;
            }
            if let Ok(number) = choice.parse::<usize>() {
                if let Some(action) = number.checked_sub(1).and_then(|index| actions.get(index)) {
                    return Flow::Value(action.clone());
                }
            }
            if let Some(action) = actions.iter().find(|action| action.id == choice) {
                return Flow::Value(action.clone());
            }
            surface.write_line("Unknown Action selection.");
        }
    }
}

fn dynamic_choices(
    input: &ActionInputDefinition,
    registry: &ActionRegistry,
    context: &ActionExecutionContext<'_>,
    surface: &mut dyn TerminalSurface,
) -> Vec<String> {
    let Some(selection) = &input.selection else {
        return Vec::new();
    };
    let Some(source) = registry.get(&selection.action) else {
        surface.write_line(&format!("Selection source Action is unavailable: {}", selection.action));
        return Vec::new();
    };
    if source.mutation_class != MutationClass::ReadOnly || source.inputs.iter().any(|candidate| candidate.required) {
        surface.write_line(&format!("Selection source Action is not safe for guided discovery: {}", selection.action));
        return Vec::new();
    }
    let result = registry.execute(&selection.action, &Value::Object(Map::new()), context);
    if !result.ok {
        let message = result.error.as_ref().map(|error| error.message.as_str()).unwrap_or("selection source failed");
        surface.write_line(&format!("Selection source unavailable: {message}"));
        return Vec::new();
    }
    let Some(items) = result
        .data
        .as_ref()
        .and_then(|data| data.get(&selection.collection))
        .and_then(Value::as_array)
    else {
        surface.write_line("Selection source did not return the declared collection.");
        return Vec::new();
    };
    items
        .iter()
        .filter_map(|item| item.get(&selection.value_field).and_then(Value::as_str).map(str::to_owned))
        .collect()
}

fn prompt_input(
    input: &ActionInputDefinition,
    registry: &ActionRegistry,
    context: &ActionExecutionContext<'_>,
    surface: &mut dyn TerminalSurface,
) -> Flow<Value> {
    let static_choices = input.choices.clone().unwrap_or_default();
    let dynamic = dynamic_choices(input, registry, context, surface);
    let displayed = if !static_choices.is_empty() { static_choices } else { dynamic };

    if !displayed.is_empty() {
        surface.write_line(&format!("Choices for {}:", input.name));
        for (index, value) in displayed.iter().enumerate() {
            surface.write_line(&format!("{}. {value}", index + 1));
        }
    }

    loop {
        let Some(raw) = surface.prompt(&format!("{} (/back, /cancel): ", input.name)) else {
            return Flow::Cancelled(cancel(None));
        };
        let value = raw.trim();
        if value == "/back" {
            return Flow::Back;
        }
        if value == "/cancel" {
            return Flow::Cancelled(cancel(None));
        }
        if value.is_empty() && input.required {
            surface.write_line("A value is required.");
            continue;
        }

        let selected = if let Ok(number) = value.parse::<usize>() {
            number.checked_sub(1).and_then(|index| displayed.get(index)).cloned()
        } else {
            None
        };
        let text = selected.unwrap_or_else(|| value.to_owned());
        if !displayed.is_empty() && !displayed.iter().any(|choice| choice == &text) {
            surface.write_line("Choose one of the available values.");
            continue;
        }

        return match input.input_type.as_str() {
            "string" => Flow::Value(Value::String(text)),
            "boolean" => match text.to_lowercase().as_str() {
                "true" | "yes" | "y" => Flow::Value(Value::Bool(true)),
                "false" | "no" | "n" => Flow::Value(Value::Bool(false)),
                _ => {
                    surface.write_line("Enter true/false or yes/no.");
                    continue;
                }
            },
            other => {
                surface.write_line(&format!("Unsupported guided input type: {other}"));
                return Flow::Cancelled(cancel(None));
            }
        };
    }
}

fn collect_input(
    action: &ActionDescriptor,
    registry: &ActionRegistry,
    context: &ActionExecutionContext<'_>,
    surface: &mut dyn TerminalSurface,
) -> Flow<Value> {
    let mut input = Map::new();
    for definition in &action.inputs {
        match prompt_input(definition, registry, context, surface) {
            Flow::Value(value) => {
                input.insert(definition.name.clone(), value);
            }
            Flow::Back => return Flow::Back,
            Flow::Cancelled(result) => return Flow::Cancelled(result),
        }
    }
    Flow::Value(Value::Object(input))
}

fn confirm_mutation(action: &ActionDescriptor, surface: &mut dyn TerminalSurface) -> Flow<()> {
    if action.mutation_class == MutationClass::ReadOnly {
        return Flow::Value(());
    }
    loop {
        let Some(raw) = surface.prompt(&format!(
            "Execute {} Action {}? [y/N] (/back, /cancel): ",
            action.mutation_class.as_str(),
            action.id,
        )) else {
            return Flow::Cancelled(cancel(Some(&action.id)));
        };
        match raw.trim().to_lowercase().as_str() {
            "y" | "yes" => return Flow::Value(()),
            "/back" => return Flow::Back,
            "/cancel" | "" | "n" | "no" => return Flow::Cancelled(cancel(Some(&action.id))),
            _ => surface.write_line("Enter yes, no, /back, or /cancel."),
        }
    }
}

pub fn run_guided_action_picker(
    registry: &ActionRegistry,
    context: &ActionExecutionContext<'_>,
    surface: &mut dyn TerminalSurface,
) -> ActionResult {
    loop {
        let action = match select_action(registry, surface) {
            Flow::Value(action) => action,
            Flow::Back => continue,
            Flow::Cancelled(result) => return result,
        };
        let input = match collect_input(&action, registry, context, surface) {
            Flow::Value(input) => input,
            Flow::Back => continue,
            Flow::Cancelled(result) => return result,
        };
        match confirm_mutation(&action, surface) {
            Flow::Value(()) => return registry.execute(&action.id, &input, context),
            Flow::Back => continue,
            Flow::Cancelled(result) => return result,
        }
    }
}
